File: examples/python/example_logistic_regression.py
#!/usr/bin/env python3
"""
Logistic Regression Example
===========================
Demonstrates binary classification using logistic regression.
"""

import machine_learning_module as ml
import numpy as np

def main():
    print("=" * 40)
    print("  LOGISTIC REGRESSION EXAMPLE")
    print("=" * 40)
    print()
    
    # 1. Generate synthetic data
    print("1. Generating synthetic classification data...")
    np.random.seed(42)
    n_samples = 1000
    
    # Class 0: centered at (1,1)
    X0 = np.random.randn(n_samples // 2, 2) + np.array([1, 1])
    y0 = np.zeros(n_samples // 2)
    
    # Class 1: centered at (3,3)
    X1 = np.random.randn(n_samples // 2, 2) + np.array([3, 3])
    y1 = np.ones(n_samples // 2)
    
    X = np.vstack([X0, X1])
    y = np.hstack([y0, y1])
    
    print(f"   Generated {n_samples} samples with 2 features")
    print("   Class 0: centered at (1,1)")
    print("   Class 1: centered at (3,3)\n")
    
    # 2. Split data
    print("2. Splitting data (80% train, 20% test)...")
    split_idx = int(0.8 * n_samples)
    X_train, X_test = X[:split_idx], X[split_idx:]
    y_train, y_test = y[:split_idx], y[split_idx:]
    print(f"   Training samples: {len(X_train)}")
    print(f"   Test samples: {len(X_test)}\n")
    
    # 3. Train model
    print("3. Training model...")
    model = ml.LogisticRegression(0.1, 1000, 0.001)
    model.fit(X_train, y_train)
    
    # 4. Evaluate
    print("\n4. Evaluation on test set:")
    accuracy = model.score(X_test, y_test)
    print(f"   Accuracy: {accuracy:.4f}")
    
    # 5. Confusion matrix
    print("\n5. Confusion Matrix:")
    cm = model.confusion_matrix(X_test, y_test)
    print("        Pred 0  Pred 1")
    print(f"  Act 0   {int(cm[0,0]):4d}   {int(cm[0,1]):4d}")
    print(f"  Act 1   {int(cm[1,0]):4d}   {int(cm[1,1]):4d}")
    
    # 6. Precision, Recall, F1
    print("\n6. Additional metrics:")
    metrics = model.precision_recall_f1(X_test, y_test)
    print(f"   Precision: {metrics[0]:.4f}")
    print(f"   Recall:    {metrics[1]:.4f}")
    print(f"   F1 Score:  {metrics[2]:.4f}")
    
    # 7. Coefficients
    print("\n7. Learned coefficients:")
    coef = model.coefficients
    intercept = model.intercept
    print(f"   Intercept: {intercept:.4f}")
    for i, c in enumerate(coef):
        print(f"   Coefficient x{i+1}: {c:.4f}")
    
    # 8. Probability predictions
    print("\n8. Probability predictions:")
    X_new = np.array([[2.0, 2.0]])
    proba = model.predict(X_new)
    print(f"   Input: x1=2.0, x2=2.0")
    print(f"   Probability of class 1: {proba[0]:.4f}")
    print(f"   Predicted class: {model.predict_class(X_new)[0]}")
    
    print("\n✅ Logistic Regression example completed!")

if __name__ == "__main__":
    main()

File: examples/cpp/example_logistic_regression.cpp
/**
 * @file logistic_regression_example.cpp
 * @brief Example of Logistic Regression usage
 * 
 * Demonstrates:
 * - Binary classification
 * - Training with regularization
 * - Confusion matrix and metrics
 */

#include <iostream>
#include <iomanip>
#include "models/logistic_regression.h"
#include "utils/math_utils.h"

using namespace models;
using namespace utils;

int main() {
    std::cout << "========================================" << std::endl;
    std::cout << "  LOGISTIC REGRESSION EXAMPLE" << std::endl;
    std::cout << "========================================\n" << std::endl;
    
    // Step 1: Generate synthetic binary classification data
    std::cout << "1. Generating synthetic classification data..." << std::endl;
    
    int n_samples = 1000;
    int n_features = 2;
    
    Eigen::MatrixXd X(n_samples, n_features);
    Eigen::VectorXd y(n_samples);
    
    std::random_device rd;
    std::mt19937 gen(42);
    std::normal_distribution<> dist(0.0, 1.0);
    
    // Class 0: centered at (1, 1)
    // Class 1: centered at (3, 3)
    for (int i = 0; i < n_samples; ++i) {
        if (i < n_samples / 2) {
            X(i, 0) = 1.0 + dist(gen);
            X(i, 1) = 1.0 + dist(gen);
            y(i) = 0;
        } else {
            X(i, 0) = 3.0 + dist(gen);
            X(i, 1) = 3.0 + dist(gen);
            y(i) = 1;
        }
    }
    
    std::cout << "   Generated " << n_samples << " samples with 2 features" << std::endl;
    std::cout << "   Class 0: centered at (1,1)" << std::endl;
    std::cout << "   Class 1: centered at (3,3)\n" << std::endl;
    
    // Step 2: Split data
    std::cout << "2. Splitting data (80% train, 20% test)..." << std::endl;
    auto splits = MathUtils::train_test_split(X, y, 0.2, 42);
    const auto& train = splits[0];
    const auto& test = splits[1];
    
    std::cout << "   Training samples: " << train.first.rows() << std::endl;
    std::cout << "   Test samples: " << test.first.rows() << "\n" << std::endl;
    
    // Step 3: Train model with different regularization
    std::cout << "3. Training models with different regularization..." << std::endl;
    
    std::vector<double> lambdas = {0.0, 0.001, 0.01, 0.1};
    
    std::cout << "   " << std::setw(15) << "Lambda"
              << std::setw(15) << "Accuracy"
              << std::setw(15) << "Precision"
              << std::setw(15) << "Recall"
              << std::setw(15) << "F1" << std::endl;
    std::cout << "   " << std::string(75, '-') << std::endl;
    
    for (double lambda : lambdas) {
        LogisticRegression model(0.1, 1000, lambda);
        model.fit(train.first, train.second);
        
        double accuracy = model.score(test.first, test.second);
        Eigen::Vector3d metrics = model.precision_recall_f1(test.first, test.second);
        
        std::cout << "   " << std::setw(15) << lambda
                  << std::setw(15) << std::fixed << std::setprecision(4) << accuracy
                  << std::setw(15) << metrics(0)
                  << std::setw(15) << metrics(1)
                  << std::setw(15) << metrics(2) << std::endl;
    }
    
    // Step 4: Train best model and show details
    std::cout << "\n4. Training best model (lambda = 0.001)..." << std::endl;
    LogisticRegression model(0.1, 1000, 0.001);
    model.fit(train.first, train.second);
    
    // Step 5: Confusion matrix
    std::cout << "\n5. Confusion Matrix on test set:" << std::endl;
    Eigen::MatrixXd cm = model.confusion_matrix(test.first, test.second);
    
    std::cout << "   " << std::setw(10) << ""
              << std::setw(10) << "Pred 0"
              << std::setw(10) << "Pred 1" << std::endl;
    std::cout << "   " << std::string(30, '-') << std::endl;
    std::cout << "   " << std::setw(10) << "Actual 0"
              << std::setw(10) << cm(0,0)
              << std::setw(10) << cm(0,1) << std::endl;
    std::cout << "   " << std::setw(10) << "Actual 1"
              << std::setw(10) << cm(1,0)
              << std::setw(10) << cm(1,1) << std::endl;
    
    // Step 6: Coefficients
    std::cout << "\n6. Learned coefficients:" << std::endl;
    Eigen::VectorXd coef = model.coefficients();
    double intercept = model.intercept();
    
    std::cout << "   Intercept: " << intercept << std::endl;
    for (int i = 0; i < coef.size(); ++i) {
        std::cout << "   Coefficient x" << i+1 << ": " << coef(i) << std::endl;
    }
    
    // Step 7: Make predictions with different thresholds
    std::cout << "\n7. Making predictions with different thresholds:" << std::endl;

    // CORRETTO: usa una matrice 2D (1 riga, 2 colonne)
    Eigen::MatrixXd x_new(1, 2);
    x_new << 2.0, 2.0;
    double proba = model.predict(x_new)(0);

    std::cout << "   Input: x1 = 2.0, x2 = 2.0" << std::endl;
    std::cout << "   Probability: " << proba << std::endl;
    std::cout << "   Class (threshold=0.5): " << (proba > 0.5 ? 1 : 0) << std::endl;
    std::cout << "   Class (threshold=0.7): " << (proba > 0.7 ? 1 : 0) << std::endl;
    std::cout << "   Class (threshold=0.3): " << (proba > 0.3 ? 1 : 0) << std::endl;
    
    return 0;
}